### PR TITLE
WIP Version 1.0.0

### PR DESCRIPTION
**Pull request by HofmannZ**
### Breaking changes

- Renamed `Client` to `GraphQLClient` to avoid name collision with other packages.
- Renamed `GraphqlProvider` to `GraphQLProvider` to align with new naming.
- Renamed `GraphqlConsumer` to `GraphQLConsumer` to align with new naming.
- Renamed `GQLError` to `GraphQLError` to align with new naming.
- `GraphQLClient` requires a `Link` to passed into the constructor.
- `GraphQLClient` no longer requires a `endPoint`  or `apiToken` to be passed into the constructor. Instead you can provide it to the `Link`.
- The `Query` and `Mutation` widgets are now `StreamBuilders`, there the api did change slightly.

#### Fixes / Enhancements

- Improved typing throughout the library.
- Queries are handled as streams of operations.
- Added the `HttpLink` to handle requests using http.
- `HttpLink` allows headers to be customised.
- The api allows contributors to write their own custom links.
- `GraphQLClient` now reads and writes data from the `Cache` based on the provided `FetchPolicy` option.
- Implemented caching for data from `FetchResults`.
- Added a document parser.
- The library now tagets Dart version `>=2.0.0 <3.0.0` as recomended by Flutter `0.6.0`.
- Removed the old client from the library.

#### Docs

- Document the new link system.
- Write an upgrade guide.
- Implement the new link system in the example.

### CODE REVIEW

File: example/lib/main.dart [hunk 1]
```diff
@@ -52,40 +59,38 @@ class _MyHomePageState extends State<MyHomePage> {
...
-        builder: ({
-          bool loading,
-          Map data,
-          Exception error,
-        }) {
-          if (error != null) {
-            return Text(error.toString());
+        options: QueryOptions(
+          document: queries.readRepositories,
+          pollInterval: 4,
```
[the remark this answers is not in the document]
Comment (HofmannZ) [reply example/lib/main.dart]: We could do that, how do you propose we would make the `Duration` from the float?

File: lib/src/core/query_manager.dart [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    // execute the operation trough the provided link(s)
+    fetchResult = await execute(
+      link: link,
+      operation: operation,
+    ).first;
+
+    // save the data from fetchResult to the cache
+    if (fetchResult.data != null) {
```
[the remark this answers is not in the document]
Comment (HofmannZ) [reply lib/src/core/query_manager.dart]: Don’t return of `FetchPolicy.no_cache` and throw exception if no data and `FetchPolicy..network_only`.

File: lib/src/cache/in_memory.dart [hunk 3]
```diff
@@ -5,9 +5,33 @@ import 'dart:io';
...
 import 'package:path_provider/path_provider.dart';
 
-class InMemoryCache {
+import 'package:graphql_flutter/src/cache/cache.dart';
+
+class InMemoryCache implements Cache {
   HashMap<String, dynamic> _inMemoryCache = HashMap<String, dynamic>();
 
+  dynamic read(String key) {
+    if (_inMemoryCache.containsKey(key)) {
```
[the remark this answers is not in the document]
Comment (HofmannZ) [reply lib/src/cache/in_memory.dart]: Good catch, also might be interesting to change dynamic to be ether a Map<String, dynamic> or List<Map<String, dynamic>>.

**Comment by smkhalsa** [comment]
@HofmannZ it looks like there hasn't been much activity on this in a while. Can you provide a status update on the project? Is it still being actively developed? Any guidance on the roadmap?

**Comment by eusdima** [comment]
@smkhalsa both me and @HofmannZ are currently working full-time, but we are going to start investing more time soon into the library, and we will let you guys know what is planned for the coming future.

Sneak peek:
- I am personally working on a schema parser that generates dart types + data validation.
- As a side project I am also working on a http scheduler to improve the smoothness on high latency, low download speed networks for the new client.

**Comment by smkhalsa** [comment]
@eusdima thanks for the update and for all the work you guys have put into the library. I very much look forward to the schema parser / Dart type generator.

Are you guys using the library internally in production?

**Comment by eusdima** [comment]
@smkhalsa we used it internally to create both Zino Sell and Zino Scan. We are creating a team of flutter engineers based in Amsterdam, so expect some more features in the coming weeks/months.

**Comment by mogol** [comment]
@eusdima 
do you have any concerns to switch from [Client](https://github.com/dart-lang/http) to [HttpClient](https://api.dartlang.org/stable/2.1.0/dart-io/HttpClient-class.html) in [HttpLink](https://github.com/zino-app/graphql-flutter/pull/76/files#diff-90b636798bea2e6558abc48672e7700f) ?
HttpClient provides more APIs to tune behaviour. E.g. it supports setting proxy and can be used with CharlesProxy.

**Pull request merged**